src/buttons.rs

```diff
@@ -0,0 +1,198 @@
+        x.enable_interrupt(&mut eic);
+        y.enable_interrupt(&mut eic);
+        z.enable_interrupt(&mut eic);
+        u.enable_interrupt(&mut eic);
+        b.enable_interrupt(&mut eic);
+
+        ButtonController {
+            _eic: eic.finalize(),
+            // b1,
+            b2,
+            b3,
+            x,
+            y,
+            z,
+            u,
+            b,
+        }
+    }
+}
+
+pub enum Button {
+    TopLeft,
+    TopMiddle,
+    // TopRight,
+    Down,
+    Up,
+    Left,
+    Right,
+    Click,
+}
+
+pub struct ButtonEvent {
+    pub button: Button,
+    pub down: bool,
+}
+
+pub struct ButtonController {
+    _eic: eic::EIC,
+    // b1: ExtInt10<Pc26<PfA>>,
+    b2: ExtInt11<Pc27<PfA>>,
+    b3: ExtInt12<Pc28<PfA>>,
+
+    x: ExtInt3<Pd8<PfA>>,
+    y: ExtInt4<Pd9<PfA>>,
+    z: ExtInt5<Pd10<PfA>>,
+    u: ExtInt10<Pd20<PfA>>,
+    b: ExtInt7<Pd12<PfA>>,
+}
+
+macro_rules! isr {
+    ($Handler:ident, $($Event:expr, $Button:ident),+) => {
+        pub fn $Handler(&mut self) -> Option<ButtonEvent> {
+            $(
+                {
+                    let b = &mut self.$Button;
+                    if b.is_interrupt() {
+                        b.clear_interrupt();
+                        return Some(ButtonEvent {
+                            button: $Event,
+                            down: !b.state(),
+                        })
+                    }
+                }
+            )+
+
+            None
+        }
+    };
+}
+
+impl ButtonController {
+    pub fn enable(&self, nvic: &mut NVIC) {
+        unsafe {
+            nvic.set_priority(interrupt::EIC_EXTINT_10, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_10);
+            nvic.set_priority(interrupt::EIC_EXTINT_11, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_11);
+            nvic.set_priority(interrupt::EIC_EXTINT_12, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_12);
+            nvic.set_priority(interrupt::EIC_EXTINT_3, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_3);
+            nvic.set_priority(interrupt::EIC_EXTINT_4, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_4);
+            nvic.set_priority(interrupt::EIC_EXTINT_5, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_5);
+            nvic.set_priority(interrupt::EIC_EXTINT_7, 1);
+            NVIC::unmask(interrupt::EIC_EXTINT_7);
+        }
+    }
+
+    isr!(interrupt_extint3, Button::Down, x);
+    isr!(interrupt_extint4, Button::Right, y);
+    isr!(interrupt_extint5, Button::Click, z);
+    isr!(interrupt_extint7, Button::Left, b);
+    isr!(interrupt_extint10, Button::Up, u);
+    // isr!(interrupt_extint10, Button::TopRight, b1);
+    isr!(interrupt_extint11, Button::TopMiddle, b2);
+    isr!(interrupt_extint12, Button::TopLeft, b3);
+}
+
+#[macro_export]
+macro_rules! button_interrupt {
+    ($controller:ident, unsafe fn $func_name:ident ($cs:ident $cstype:ty, $event:ident ButtonEvent ) $code:block) => {
```

Comment:
```suggestion
    ($controller:ident, unsafe fn $func_name:ident ($cs:ident: $cstype:ty, $event:ident: ButtonEvent ) $code:block) => {
```
I think requiring the colons here is probably better, as in the macro invocation it then just looks like a regular function. I initially thought the buttons example contained an error when I noticed them missing. What are your thoughts on this?

Reply:
Whoops, I was in golang mode with the function syntax. Fixed.